cognite/client/data_classes/datapoints.py

```diff
@@ -189,15 +204,54 @@ def to_pandas(self, column_names="externalId") -> "pandas.DataFrame":
             identifier = self.id
         else:
             raise ValueError("column_names must be 'externalId' or 'id'")
-        for attr, value in self._get_non_empty_data_fields(get_empty_lists=True):
+        fields = self._get_non_empty_data_fields(get_empty_lists=True)
+        for attr, value in fields:
             if attr == "timestamp":
                 timestamps = value
             else:
                 id_with_agg = str(identifier)
                 if attr != "value":
-                    id_with_agg += "|{}".format(utils._auxiliary.to_camel_case(attr))
+                    id_with_agg = "{}|{}".format(id_with_agg, utils._auxiliary.to_camel_case(attr))
                 data_fields[id_with_agg] = value
-        return pd.DataFrame(data_fields, index=pd.DatetimeIndex(data=np.array(timestamps, dtype="datetime64[ms]")))
+
+        df = pd.DataFrame(data_fields, index=pd.DatetimeIndex(data=np.array(timestamps, dtype="datetime64[ms]")))
+        if complete and df.shape[0] > 1:
+            if complete_granularity:
+                ms_granularity = cognite.client.utils._time.granularity_to_ms(complete_granularity)
+            else:
+                ms_granularity = np.diff(timestamps).min()
+            bounds = np.array(
+                [complete_start if complete_start is not None else df.index[0], complete_end or df.index[-1]]
+            ).astype("datetime64[ms]")
+            df = df.reindex(
+                np.arange(
+                    bounds[0],
+                    bounds[1] + pd.Timedelta(microseconds=1),
+                    pd.Timedelta(microseconds=ms_granularity * 1000),
+                ),
+                copy=False,
+            )
+
+            df.fillna({c: 0 for c in df.columns if re.search(c, r"\|(sum|totalVariance|count)$")}, inplace=True)
+            lin_int_cols = [c for c in df.columns if re.search(c, r"\|interpolation$")] if not complete_is_step else []
+            step_int_cols = [
+                c
+                for c in df.columns
+                if re.search(c, r"\|stepInterpolation$") or (complete_is_step and re.search(c, r"\|interpolation$"))
+            ]
+            df[lin_int_cols] = df[lin_int_cols].interpolate(limit_area="inside")
+            df[step_int_cols] = df[step_int_cols].ffill()
+
+        if not include_aggregate_name:
+            if len(fields) > 1:
+                raise ValueError(
+                    "include_aggregate_name=False can not be used here as it would lead to duplicate column names for multiple aggregates {}".format(
+                        fields
+                    )
+                )
+            df.rename(columns=lambda s: re.sub(s, r"\|\w+$", ""), inplace=True)
+
+        return df
```

Comment:
This is a very long method. Maybe try splitting it up so it is a little more readable?

Reply:
done